Combine and add expression tree restrictions (#36991)
dotnet/docs: docs/csharp/language-reference/compiler-messages/expression-tree-restrictions.md

- **CS8640** - *Expression tree cannot contain value of ref struct or restricted type.*
- **CS8642** - *An expression tree may not contain a null coalescing assignment.*
- **CS8790** - *An expression tree may not contain a pattern <xref:System.Index?displayProperty=fullName> or <xref:System.Range?displayProperty=fullName> indexer access.*
- **CS8791** - *An expression tree may not contain a from-end index ('`^`') expression.*
- **CS8792** - *An expression tree may not contain a range ('`..`') expression.*
- **CS8810** - *'`&`' on method groups cannot be used in expression trees.*
- **CS8849** - *An expression tree may not contain a `with`-expression.*
- **CS8927** - *An expression tree may not contain an access of static virtual or abstract interface member.*
- **CS8952** - *An expression tree may not contain an interpolated string handler conversion.*
- **CS8972** - *A lambda expression with attributes cannot be converted to an expression tree.*
- **CS9170** - *An expression tree may not contain an inline array access or conversion.*
- **CS9175** - *An expression tree may not contain a collection expression.*

## Expression tree restrictions

> All of the errors in the preceding list indicate you've used a C# expression type that isn't allowed in an expression tree. In most cases, the prohibited expressions represent syntax introduced after C# 3.0. These expressions are prohibited because allowing them would create a breaking change in all libraries that parse expression trees. Without updates, those libraries would all require updates before supporting any new expression trees.

Reviewer: I think this is redundant, but I'm not totally sure what it's trying to explain? It's the last sentence.

> Without updates, those libraries would all require updates before supporting any new expression trees.

> All of the errors in the preceding list indicate you've used a C# expression type that isn't allowed in an expression tree. In most cases, the prohibited expressions represent syntax introduced after C# 3.0. These expressions are prohibited because allowing them would create a breaking change in all libraries that parse expression trees. Without updates, those libraries would all require updates before supporting any new expression trees.

Author: reword last sentence:

```suggestion
All of the errors in the preceding list indicate you've used a C# expression type that isn't allowed in an expression tree. In most cases, the prohibited expressions represent syntax introduced after C# 3.0. These expressions are prohibited because allowing them would create a breaking change in all libraries that parse expression trees. All libraries would need to be enhanced to parse new C# expressions if newer constructs were allowed.
```

- [assignment](../operators/assignment-operator.md) operations.
- [statement lambdas](../operators/lambda-expressions.md#statement-lambdas) aren't allowed.
- [multi-dimensional array](../builtin-types/arrays.md#multidimensional-arrays) initializers. Instead, you must create and initialize a multi-dimensional array outside of the expression tree.
- [`dynamic`](../builtin-types/reference-types.md#the-dynamic-type) operations aren't allowed.
- [pattern matching](../operators/patterns.md) expressions aren't allowed.
- [Tuple literals](../builtin-types/value-tuples.md) and many tuple operations, such as equality comparisons aren't allowed.
- [`throw` expressions](../statements/exception-handling-statements.md#the-throw-expression) aren't allowed.
- [discard](../../fundamentals/functional/discards.md) (`_`) declarations.
- The [index and range](../operators/member-access-operators.md#indexer-access) operators aren't allowed.
- Non-destructive mutation using [`with`](../operators/with-expression.md) expressions aren't allowed.
- You can't declare or access [inline arrays](../builtin-types/struct.md#inline-arrays).

> - You can't include [Collection expressions](../operators/collection-expressions.md).

Reviewer: I don't think this needs to be capitalized.
```suggestion
- You can't include [collection expressions](../operators/collection-expressions.md).
```

- `ref` returns aren't allowed.
- Calls to methods that return by `ref` aren't allowed.
- [static abstract interface members](../keywords/interface.md#static-abstract-and-virtual-members) can't be accessed.

Other restrictions are:

- Attributes can't be applied to the lambda expression, its parameters or return.
- The lambda expression must be convertible to a type derived from <xref:System.Linq.Expressions.Expression?displayProperty=fullName> whose type parameter is a delegate type.
- [named and optional parameters](../../programming-guide/classes-and-structs/named-and-optional-arguments.md) are restricted. The expression can't call a method specifying named arguments, and it can't use the default value of an optional parameter.
- [Dictionary initializers](../../programming-guide/classes-and-structs/object-and-collection-initializers.md#collection-initializers) aren't allowed. Neither are extension `Add` methods.
- The target expression must be a lambda expression. Constants and variables aren't allowed, but a lambda expression that returns a constant or variable is.
- Unsafe pointer operations aren't allowed.
- COM calls must include `ref` on arguments; it can't be implied.

> - The unsupported `__arglist` keyword is not allowed.

Reviewer: acro
```suggestion
- The unsupported `__arglist` keyword isn't allowed.
```
